pkg/cmd/roachtest/tests/import_cancellation.go

```diff
@@ -0,0 +1,307 @@
+}
+
+func getRangeKeySetsTotal(ctx context.Context, c cluster.Cluster, t test.Test) int {
+	adminURLs, err := c.ExternalAdminUIAddr(ctx, t.L(), c.Node(1))
+	if err != nil {
+		t.Fatal(err)
+	}
+	now := timeutil.Now()
+	start := now.Add(-60 * time.Second)
+	response := mustGetMetrics(t, adminURLs[0], start, now, []tsQuery{
+		{
+			name:      "cr.node.storage.keys.range-key-set.count",
+			queryType: total,
+		},
+	})
+	if len(response.Results[0].Datapoints) <= 1 {
+		t.Fatalf("not enough datapoints in timeseries query response: %+v", response)
+	}
+	datapoints := response.Results[0].Datapoints
+	return int(datapoints[len(datapoints)-1].Value)
+}
+
+type importCancellationTest struct {
+	test.Test
+	c       cluster.Cluster
+	rootRng *rand.Rand
+	seed    int64
+}
+
+func (t *importCancellationTest) makeFilename(tableName string, number int, numFiles int) string {
+	// Tables with more than one files have the number as a suffix on the
+	// filename, `<tablename>.tbl.1`. Tables with a single file do not.
+	if numFiles > 1 {
+		return fmt.Sprintf(`'gs://cockroach-fixtures/tpch-csv/sf-100/%[1]s.tbl.%[2]d?AUTH=implicit'`, tableName, number)
+	}
+	return fmt.Sprintf(`'gs://cockroach-fixtures/tpch-csv/sf-100/%[1]s.tbl?AUTH=implicit'`, tableName)
+}
+
+func (t *importCancellationTest) runImportSequence(
+	ctx context.Context, rng *rand.Rand, tableName string, numFiles int,
+) {
+	n := t.c.All()[rng.Intn(len(t.c.All()))]
+	conn := t.c.Conn(ctx, t.L(), n)
+	if _, err := conn.Exec(`USE csv;`); err != nil {
+		t.Fatal(err)
+	}
+
+	filesToImport := make([]string, numFiles)
+	for i := 0; i < numFiles; i++ {
+		filesToImport[i] = t.makeFilename(tableName, i+1, numFiles)
+	}
+
+	attempts := randutil.RandIntInRange(rng, 2, 5)
+	for i := 0; i < attempts && len(filesToImport) > 0; i++ {
+		t.WorkerStatus(fmt.Sprintf(`attempt %d/%d for table %q`, i+1, attempts, tableName))
+		files := filesToImport
+		// If not the last attempt, import a subset of the files available.
+		// This will create MVCC range tombstones across separate regions of
+		// the table's keyspace.
+		if i != attempts-1 {
+			rng.Shuffle(len(files), func(i, j int) {
+				files[i], files[j] = files[j], files[i]
+			})
+			files = files[:randutil.RandIntInRange(rng, 1, len(files)+1)]
+		}
+		t.L().PrintfCtx(ctx, "Beginning import job for attempt %d/%d for table %s with %d files: %s",
+			i+1, attempts, tableName, len(files), strings.Join(files, ", "))
+
+		var jobID string
+		importStmt := fmt.Sprintf(`
+			IMPORT INTO %[1]s
+			CSV DATA (
+			%[2]s
+			) WITH  delimiter='|', detached
+		`, tableName, strings.Join(files, ", "))
+		err := conn.QueryRowContext(ctx, importStmt).Scan(&jobID)
+		if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+			t.Fatal(err)
+		}
+
+		// If not the last attempt, cancel the job after a random duration in [10s,3m).
+		if i != attempts-1 {
+			timeout := time.Duration(randutil.RandIntInRange(rng, 10, 60*3)) * time.Second
```

Comment:
The import may complete before this duration, which I think makes this test prone to flakes or unexpected behavior. For example, an IMPORT rollback on a table that had data pre-import uses "predicate based" delete range (issues some point and range tombstones) while a rollback without pre-existing data issues one big range tombstone. (aside: Given this, I wonder if we should have two roachtests: one testing import cancellations on empty tables, and one for non-empty tables).

To ensure the import doesn't complete, you can automatically pause the import job after all data ingestion by setting the `import.after_ingest` jobs pause point. via

```
SET CLUSTER SETTING jobs.debug.pausepoints = 'import.after_ingest';
```

note that when you _want_ the job to complete, set the value of this cluster setting to `''` before starting the import.